libp2p/NodeTable.h

```diff
@@ -0,0 +1,354 @@
+ *
+ * Thread-safety is ensured by modifying NodeEntry details via 
+ * shared_ptr replacement instead of mutating values.
+ *
+ * [Integration]
+ * @todo deadline-timer which maintains tcp/peer connections
+ * @todo restore nodes: affects refreshbuckets
+ * @todo TCP endpoints
+ * @todo makeRequired: don't try to evict node if node isRequired.
+ * @todo makeRequired: exclude bucket from refresh if we have node as peer.
+ *
+ * [Optimization]
+ * @todo Pong to include ip:port where ping was received
+ * @todo expiration and sha3(id) 'to' for messages which are replies (prevents replay)
+ * @todo std::shared_ptr<PingNode> m_cachedPingPacket;
+ * @todo std::shared_ptr<FindNeighbors> m_cachedFindSelfPacket;
+ * @todo store root node in table?
+ *
+ * [Networking]
+ * @todo TCP endpoints
+ * @todo eth/upnp/natpmp/stun/ice/etc for public-discovery
+ * @todo firewall
+ *
+ * [Protocol]
+ * @todo post-eviction pong
+ * @todo optimize knowledge at opposite edges; eg, s_bitsPerStep lookups. (Can be done via pointers to NodeBucket)
+ * @todo ^ s_bitsPerStep = 5; // Denoted by b in [Kademlia]. Bits by which address space is divided.
+ * @todo optimize (use tree for state and/or custom compare for cache)
+ * @todo reputation (aka universal siblings lists)
+ */
+class NodeTable: UDPSocketEvents, public std::enable_shared_from_this<NodeTable>
+{
+	friend struct Neighbors;
+	using NodeSocket = UDPSocket<NodeTable, 1280>;
+	using TimePoint = std::chrono::steady_clock::time_point;
+	using EvictionTimeout = std::pair<std::pair<NodeId,TimePoint>,NodeId>;	///< First NodeId may be evicted and replaced with second NodeId.
+
+	struct NodeDefaultEndpoint
+	{
+		NodeDefaultEndpoint(bi::udp::endpoint _udp): udp(_udp) {}
+		bi::udp::endpoint udp;
+	};
+	
+	struct Node
+	{
+		Node(Public _pubk, NodeDefaultEndpoint _udp): id(_pubk), endpoint(_udp) {}
+		Node(Public _pubk, bi::udp::endpoint _udp): Node(_pubk, NodeDefaultEndpoint(_udp)) {}
+		
+		virtual NodeId const& address() const { return id; }
+		virtual Public const& publicKey() const { return id; }
+		
+		NodeId id;
+		NodeDefaultEndpoint endpoint;
+	};
+	
+	/**
+	 * NodeEntry
+	 * @todo Type of id will become template parameter.
+	 */
+	struct NodeEntry: public Node
+	{
+		NodeEntry(Node _src, Public _pubk, NodeDefaultEndpoint _gw): Node(_pubk, _gw), distance(dist(_src.id,_pubk)) {}
+		NodeEntry(Node _src, Public _pubk, bi::udp::endpoint _udp): Node(_pubk, NodeDefaultEndpoint(_udp)), distance(dist(_src.id,_pubk)) {}
+
+		const unsigned distance;	///< Node's distance from _src (see constructor).
+	};
+	
+	struct NodeBucket
+	{
+		unsigned distance;
+		TimePoint modified;
+		std::list<std::weak_ptr<NodeEntry>> nodes;
+	};
+	
+public:
+	NodeTable(ba::io_service& _io, KeyPair _alias, uint16_t _port = 30300);
+	~NodeTable();
+	
+	/// Constants for Kademlia, mostly derived from address space.
+	
+	static unsigned const s_addressByteSize = sizeof(NodeId);				///< Size of address type in bytes.
```

Comment:
Ah, other parts of code are doing this now. Very happy to do the same. Next PR :)